R/newton_solver.R

```diff
@@ -54,12 +54,7 @@ solve_final_size_newton <- function(contact_matrix,
   dx_f <- function(x) {
     cache_m <- f2(x)
     cache <- -f1(x)
-    # partial pivoting LU decomposition
-    cache_m_pivlu <- Matrix::lu(cache_m)
-    cache_m_pivlu <- Matrix::expand(cache_m_pivlu)
-    cache_m_pivlu <- cache_m_pivlu$L %*% cache_m_pivlu$U
```

Comment:
I initially added this step when the Newton solver was failing at particular `r0` (~2, 4, 12), because (partial?) pivoting LU decomposition is used in Edwin's code: https://gitlab.com/epidemics-r/code_snippets/-/blob/feature/newton_solver/include/finalsize.hpp#L166, and using `Matrix::lu` appeared to be the correct way to do this in R.

After finding that the implementation of contact matrix filling (your comment above) did not impact the results, I looked to see what other code might be redundant. This chunk seemed suspicious, and the full test suite (including language and solver equivalence) passed after removing it, so I did. (I've also completed a Julia implementation, which doesn't require this pivoting LU decomp either). Overall the evidence points to this step being unnecessary except for `Eigen`'s `solve`. Once again, I don't have the theoretical background to guess why this is the case.

You're correct about the speed gain, of course, see this benchmarking from this morning from a conversation with Tim. The Newton solver is now takes about 2x time of the iterative solver in R, whereas it was about 5x before this step was removed.

```r
library(finalsize)
# prepare arguments
contact_matrix <- c(
  5.329620e-08, 1.321156e-08, 1.832293e-08, 7.743492e-09, 5.888440e-09,
  2.267918e-09, 1.321156e-08, 4.662496e-08, 1.574182e-08, 1.510582e-08,
  7.943038e-09, 3.324235e-09, 1.832293e-08, 1.574182e-08, 2.331416e-08,
  1.586565e-08, 1.146566e-08, 5.993247e-09, 7.743492e-09, 1.510582e-08,
  1.586565e-08, 2.038011e-08, 1.221124e-08, 9.049331e-09, 5.888440e-09,
  7.943038e-09, 1.146566e-08, 1.221124e-08, 1.545822e-08, 8.106812e-09,
  2.267918e-09, 3.324235e-09, 5.993247e-09, 9.049331e-09, 8.106812e-09,
  1.572736e-08
) |> matrix(6, 6)

# make a demography vector
demography_vector <- c(
  10831795, 11612456, 13511496,
  11499398, 8167102, 4587765
)
# get an example r0
r0 <- 1.3
contact_matrix = r0 * contact_matrix

# a p_susceptibility matrix
p_susceptibility <- matrix(0.5, nrow(contact_matrix), 2)
susceptibility <- matrix(c(0.1, 0.7), nrow(contact_matrix), 2, byrow = T)

# benchmark implementations
microbenchmark::microbenchmark(
  times = 1000,
  "iterative_solver_r" = final_size_grps(
    contact_matrix = contact_matrix,
    demography_vector = demography_vector,
    susceptibility = susceptibility,
    p_susceptibility = p_susceptibility,
    solver = "iterative",
    control = list(
      iterations = 10000,
      tolerance = 1e-6
    )
  ),
  "newton_solver_r" = final_size_grps(
    contact_matrix = contact_matrix,
    demography_vector = demography_vector,
    susceptibility = susceptibility,
    p_susceptibility = p_susceptibility,
    solver = "newton",
    control = list(
      iterations = 10000,
      tolerance = 1e-6
    )
  ),
  "iterative_solver_cpp" = final_size_grps_cpp(
    contact_matrix = contact_matrix,
    demography_vector = demography_vector,
    susceptibility = susceptibility,
    p_susceptibility = p_susceptibility,
    solver = "iterative",
    control = list(
      iterations = 10000,
      tolerance = 1e-6,
      step_rate = 1.9,
      adapt_step = TRUE
    )
  ),
  "newton_solver_cpp" = final_size_grps_cpp(
    contact_matrix = contact_matrix,
    demography_vector = demography_vector,
    susceptibility = susceptibility,
    p_susceptibility = p_susceptibility,
    solver = "newton",
    control = list(
      iterations = 10000,
      tolerance = 1e-6
    )
  )
)
#> Warning in microbenchmark::microbenchmark(times = 1000, iterative_solver_r =
#> final_size_grps(contact_matrix = contact_matrix, : less accurate nanosecond
#> times to avoid potential integer overflows
#> Unit: microseconds
#>                  expr     min       lq       mean  median       uq      max
#>    iterative_solver_r  63.509  67.7730  75.760374  69.577  74.3535 4012.588
#>       newton_solver_r 136.653 145.2425 166.597596 148.789 156.4970 5910.601
#>  iterative_solver_cpp   5.371   5.8630   6.413671   6.273   6.6420   15.170
#>     newton_solver_cpp  12.628  13.3660  14.171486  13.858  14.5140   53.628
#>  neval
#>   1000
#>   1000
#>   1000
#>   1000
```

<sup>Created on 2022-10-12 by the [reprex package](https://reprex.tidyverse.org/) (v2.0.1)</sup>